gossip/minimal/goshawk.go

```diff
@@ -56,6 +61,132 @@ type FetchOptions struct {
+	// GetHubIndex returns the index at which scanning should commence for the given hub URL.
+	GetHubIndex(hubURL string) int64
+	// UpdateHubIndex indicates that future scanning of the given hub should start from at least nextIndex.
+	UpdateHubIndex(hubURL string, nextIndex int64)
+	// Flush ensures state is stored.
+	Flush(ctx context.Context) error
+}
+
+// ScanState holds a (mutex-protected) map of hub URL to next index to scan from.
+type ScanState struct {
+	Mu   sync.Mutex
+	Next map[string]int64
+}
+
+// GetHubIndex returns the index at which scanning should commence for the given hub URL.
+func (s *ScanState) GetHubIndex(hubURL string) int64 {
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+	return s.Next[hubURL]
+}
+
+// UpdateHubIndex indicates that future scanning of the given hub should start from at least nextIndex.
+func (s *ScanState) UpdateHubIndex(hubURL string, nextIndex int64) {
+	s.Mu.Lock()
+	defer s.Mu.Unlock()
+	cur := s.Next[hubURL]
+	if nextIndex > cur {
+		s.Next[hubURL] = nextIndex
+	}
+}
+
+// memoryStateManager is an implementation of the ScanStateManager interface
+// that does not persist any data; it is intended for use as a fallback and in
+// testing.
+type memoryStateManager struct {
+	ScanState
+}
+
+// Flush does nothing for a memoryStateManager.
+func (m *memoryStateManager) Flush(ctx context.Context) error {
+	// No persistent storage.
+	return nil
+}
+
+// FileStateManager is an implementation of the ScanStateManager interface
+// that stores the scan state in a writable flat file.
+type FileStateManager struct {
+	ScanState
+	filename string
+}
+
+// NewFileStateManager creates a FileStateManager from the contents of the given
+// filename, which can be be empty but should be over-writable.
+func NewFileStateManager(filename string) (*FileStateManager, error) {
+	f := FileStateManager{
+		ScanState: ScanState{Next: make(map[string]int64)},
+		filename:  filename,
+	}
+
+	file, err := os.Open(filename)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to open file for initial read: %v", err)
+	}
+	if err == nil {
+		scanner := bufio.NewScanner(file)
+		line := 0
+		for scanner.Scan() {
+			line++
+			fields := strings.Split(scanner.Text(), "\t")
+			if len(fields) != 2 {
+				return nil, fmt.Errorf("%s:%d: found unexpected number of fields", filename, line)
+			}
+			index, err := strconv.ParseInt(fields[1], 10, 64)
+			if err != nil {
+				return nil, fmt.Errorf("%s:%d: failed to parse index: %v", filename, line, err)
+			}
+			f.ScanState.Next[fields[0]] = index
+		}
+		file.Close()
+	}
+
+	// Check we can immediately write the same contents back.
+	if err := f.flush(); err != nil {
+		return nil, fmt.Errorf("%s: failed to write contents: %v", filename, err)
+	}
+
+	return &f, nil
+}
+
+func (f *FileStateManager) flush() error {
+	keys := make([]string, 0, len(f.ScanState.Next))
+	for key := range f.ScanState.Next {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
```

Comment:
Is sorting buying us anything?

Reply:
Hmm, I'm sure I had a reason for doing this, but I can't now remember what it was.  (Note to self: write more comments)

Reply:
Removed.